Implemented geometric transformation of images

vikasTmz opened the pull request:
Introduced enhancement #146 

@fspindle The remaining pixels around the transformed image now contain random noise after the interpolation. What do I do about it? refer : `line: 501`

fspindle commented:
Your code is not easy to understand.

Could you please update the header documentation of the class to introduce doxygen equations that may help to understand the algorithm.

You can also give more info here to explain a little bit more in depth.

I'm also a little bit confused by your input matrix that is a 3x3 with the translation in the 3rd row.

vikasTmz commented:
I've expanded the header documentation. Do point out if there are format/syntax issues since I am new to doxygen.

Most of the theory that I have used is from this [Lecture](http://eeweb.poly.edu/~yao/EL5123/lecture12_ImageWarping.pdf) (_refer pg 23-35_)

>I'm also a little bit confused by your input matrix that is a 3x3 with the translation in the 3rd row.

This is explained in the header documentation. I have tried to follow the convention used in other libraries that allow Image warping given transformation matrix as input.

The code is probably a bit confusing to understand as : 
- I couldn't find an alternative to `meshgrid`.
- `B` argument in `solveBySVD` is a column vector. Since I wanted x and B in `Ax = B` to be matrices (basically solve multiple linear equations), had to breakdown B into `ColVectors`. 
- Combined the idea of `getAffineTransform()` and `warpAffine()` defined by OpenCV into one function.

I will try to simplify the current implementation based on further input.

vikasTmz commented:
@fspindle  updates?

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -57,12 +57,15 @@
 #include <math.h>
 #include <string.h>
 
+
```
Remove blank line

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -409,6 +417,119 @@ void vpImageTools::crop(const unsigned char *bitmap, unsigned int width, unsigne
+  float max_2 = std::max(transcp[1][0],std::max(transcp[1][1],std::max(transcp[1][2],std::max(transcp[1][3],transcp[1][4])))); ///< max(transcp(1,:)) 
+  float min_2 = std::min(transcp[1][0],std::min(transcp[1][1],std::min(transcp[1][2],std::min(transcp[1][3],transcp[1][4])))); ///< min(transcp(1,:)) 
+
+
```
Remove blank line

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,130 @@
+#include <visp3/core/vpIoTools.h>
+#include <visp3/core/vpImageTools.h>
+
+
```
Remove 1 blank line

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,130 @@
+    if (!opt_ipath.empty()) {
+      ipath = opt_ipath;
+    }
+
```
remove 1 blank line

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,130 @@
+
+    I = I_warped_Klimt;
+    #if defined(VISP_HAVE_X11)
+        vpDisplayX d(I);
```
From line 105 to 121, indent code

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -409,6 +417,119 @@ void vpImageTools::crop(const unsigned char *bitmap, unsigned int width, unsigne
+
+  transcp = transform_matrix * transcp;
+
+  float max_1 = std::max(transcp[0][0],std::max(transcp[0][1],std::max(transcp[0][2],std::max(transcp[0][3],transcp[0][4])))); ///< max(transcp(0,:)) 
```
Introduce a space after the comma in the 4 next lines

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -409,6 +417,119 @@ void vpImageTools::crop(const unsigned char *bitmap, unsigned int width, unsigne
+{
+
+  vpMatrix transcp(3,5);
+  transcp[0][0] = 1; transcp[0][1] = 1;  transcp[0][2] = 1;             transcp[0][3] = I.getWidth();  transcp[0][4] = I.getWidth();
```
Align all the transcp[][3] and transcp[][4]

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -155,6 +158,11 @@ class VISP_EXPORT vpImageTools
 
+  template<class Type>
+  static void warpAffine(vpImage<Type> &I,
+                        vpMatrix &transform_matrix);
```
transform_matrix should be const

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -409,6 +417,119 @@ void vpImageTools::crop(const unsigned char *bitmap, unsigned int width, unsigne
+*/
+template<class Type>
+void vpImageTools::warpAffine(vpImage<Type> &I,
+                              vpMatrix &transform_matrix)
```
rename `transformation_matrix` in `T` to match the equations

fspindle commented:
- please take into account my previous reviews to improve the code
- modify the prototype of the function like
````
template<class Type>
void vpImageTools::warpAffine(const vpImage<Type> &I, 
                              const vpMatrix &transform_matrix, 
                              vpMatrix<Type> &Iwarp)

````
- background should be set to 0 (black) 
- I improved the doxygen doc. Just copy past the following:
````
/*!

  Apply an affine transformation to an image.

  The equation to warp pixel \f$(u,v)\f$ to \f$(x,y)\f$ using an affine
  transformation is the following:

  \f[
  \left\{
  \begin{array}{ccccccc}
  x & = & a_0 & + & a_1 \; u & + & a_2 \; v\\
  y & = & b_0 & + & b_1 \; u & + & b_2 \; v
  \end{array}
  \right.
  \f]

  where \f$a_0, a_1, a_2, b_0, b_1, b_2\f$ are the affine parameters.

  Let \f${\bf X} = [x \; y]^T\f$ and \f${\bf U} = [u \; v]^T\f$. Therefore the
  previous equation could be written in matrix form : \f[ {\bf X} = {\bf B} + {\bf A} \; {\bf U} \f]
  where \f${\bf A} = [ a_1 \; a_2 ; b_1 \; b_2 ]\f$ and \f${\bf B} = [a_0 \; b_0]^T\f$

  Thus the transformation matrix that is given as parameter \e T should be filled such as:

  \f[
  {\bf T} =
  \left[
  \begin{array}{cc}
  {\bf A}^T & 0 \\
  {\bf B}^T & 1
  \end{array}
  \right]
  =
  \left[
  \begin{array}{ccc}
  a_1 & b_1 & 0 \\
  a_2 & b_2 & 0 \\
  a_0 & b_0 & 1
  \end{array}
  \right]
  \f]

  To compute the transformed image inverse mapping is used over other methods like forward mapping
  as warping points are often non-integer samples:
  - For each point \f$(x,y)\f$ in the transformed image to be obtained, find its corresponding
    point \f$(u,v)\f$ in the original image.
  - If the mapped point \f$(u,v)\f$ is not an integer sample, apply interpolation using nearest
    neighbour interpolation.

  \param I : Input image to be transformed. 
  \param T : 3x3 matrix corresponding to \f${\bf T}\f$ that contains all possible geometric
  transformations, also called affine mapping.
  \param Iwarp : Resulting warped image. 

  An example that uses warpAffine() is given in testImageWarpAffine.cpp
*/
````
- to test the documentation run `make -j4 visp_doc` (need doxygen detected during ViSP configuration with cmake)
- `testWarpAffine.cpp` is failing with [travis](https://travis-ci.org/lagadic/visp/builds/213469751). You can reproduce the issue entering in `<build tree>/modules/core`and running `ctest` from command line. The issue is due to the blocking `getClick()`. The solution is nothing more than introducing the `-c` and `-d` command lines options that are called behind `ctest`. It could be also useful to introduce `-p` option to allow the user to test on an other image than `klimt.pgm`. Finally `-i` option should also be introduced as all the other examples that are using images from the data set. You can do the same as in `testImageResize.cpp` and `testImageFilter.cpp` introducing:
````
#include <visp3/io/vpParseArgv.h>
#define GETOPTARGS "cdhi:p:"

void usage(const char *name, const char *badparam, std::string ipath)
{
....
  OPTIONS:                                               Default\n\
    -i <input image path>                                %s\n\
       Set image input path.\n\
       From this path read \"ViSP-images/Klimt/Klimt.pgm\"\n\
       image.\n\
       Setting the VISP_INPUT_IMAGE_PATH environment\n\
       variable produces the same behaviour than using\n\
       this option.\n\
  \n\
    -p <personal image path>                               \n\
       Path to an image used to test image warping.\n\
       Example: -p /my_path_to/image.png\n\
  \n\
    -c                                   \n\
       Disable mouse click.\n\
  \n\
    -d                                   \n\
       Disable image display.\n\
  \n\
    -h\n\
       Print the help.\n\n";
}

bool getOptions(int argc, const char **argv, std::string &ipath,  std::string &ppath, bool &opt_display, bool &opt_click)
````
- `testImageWarpAffine.cpp` fails when I set the following translation. The final image has size 1x1
````
     // Test Translation
     transform_matrix[0][0]= 0; transform_matrix[0][1] =  0; transform_matrix[0][2] =  0;
     transform_matrix[1][0]= 0; transform_matrix[1][1] =  0; transform_matrix[1][2] =  0;
     transform_matrix[2][0]= 0.5; transform_matrix[2][1] =  0.5; transform_matrix[2][2] =  1;
````
- `testImageWarpAffine.cpp` should be extended with a translation and pure rotation example using `cos(vpMath::rad(angle))` and `sin(vpMath::rad(angle))` to fill the rotation matrix.

vikasTmz commented:
>background should be set to 0 (black)
I'm not sure on how to do this since its values are assigned during interpolation. Ideally these pixel shouldn't get any values i.e assigned to Nan, during interpolation.

vikasTmz commented:
@fspindle I've made the necessary changes : 

- generalized transformation matrix
- added doxygen doc
- set background pixels to 0 during interpolation
- added -p flag for testImageWarp
- solved travis failure
- coding style fixed

vikasTmz commented:
Hello @fspindle, updates on this? Previous comment has the summary of the changes.

vikasTmz commented:
Are there any more changes or improvements needed?

fspindle commented:
Sorry, I was busy. I will review the code asap.

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -409,6 +420,168 @@ void vpImageTools::crop(const unsigned char *bitmap, unsigned int width, unsigne
+      for (unsigned int j = 0; j < Iwarp.getCols(); j++) {
+
+        if (floor((i - T[2][1]) / Iwarp.getRows()) == 0 && floor((j - T[2][0]) / Iwarp.getCols()) == 0)
+          Iwarp[i][j] =  Iwarp_temp[i - (int)T[2][1]][j - (int)T[2][0]];
```
`floor((i - T[2][1]) / Iwarp.getRows()` should be computed just once for each row between line 566 and 567

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,288 @@
+    %s [-i <input image path>] [-p <personal image path>] [-c] [-d]\n\
+       [-h]\n                 \
+  ", name);
+
```
Could be nice to introduce 6 additional options, one for each parameter a0, a1, a2 and b0, b1, b2 that allow the user test more transformations than just the default one that is hardcoded.

vikasTmz replied:
Should these be the input parameters or could the user instead input the `angle`, `scale`(x,y) and `translation`(x,y) ? Should it throw an error if one of the parameter is missing or assume default value as 0?

fspindle replied:
We should consider command line options with more than one character. This is possible using `vpParseArgv` class

`GETOPTARGS` and usage() should be removed.
Then `getOption()` should be modified like:

```
  /*!
    Set the program options.

    \param argc : Command line number of parameters.
    \param argv : Array of command line parameters.
    \param opt_ipath: Input image path.
    \param opt_ppath: Personal image path.
    \param opt_display : Do not display if set.
    \param opt_click : Do not need click if set.
    \param opt_angle : .............
    \param opt_scale_x : .............
    \param opt_scale_y : .............
    \param opt_translation_x : .............
    \param opt_translation_y : .............
    \return false if the program has to be stopped, true otherwise.
  */
  bool getOptions(int argc, const char **argv, std::string &opt_ipath, std::string &opt_ppath,
                  bool &opt_display, bool &opt_click, float &opt_angle,
                  float &opt_scale_x, float &opt_scale_y,
                  float &opt_translation_x, float &opt_translation_y)
  {
    char *opt_ipath_ = NULL;
    char *opt_ppath_ = NULL;

    vpParseArgv::vpArgvInfo argTable[] =
    {
      {"-i", vpParseArgv::ARGV_STRING, (char*) NULL, (char *) &opt_ipath_,
       "Set image input path. From this path read \"ViSP-images/Klimt/Klimt.pgm\""},
      {"-p", vpParseArgv::ARGV_STRING, (char*) NULL, (char *) &opt_ppath_,
       "Path to an image used to test image warping."},
      {"-c", vpParseArgv::ARGV_CONSTANT, 0, (char *) &opt_click,
       "Disable mouse click."},
      {"-d", vpParseArgv::ARGV_CONSTANT, 0, (char *) &opt_display,
       "Disable display."},
      {"-angle", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_angle,
       "Angle in degrees."},
      {"-scale_x", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_scale_x,
       "Scale factor along x."},
      {"-scale_y", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_scale_y,
       "Scale factor along y."},
      {"-trans_x", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_translation_x,
       "Translation along x."},
      {"-trans_y", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_translation_y,
       "Translation along y."},
      {"-h", vpParseArgv::ARGV_HELP, (char*) NULL, (char *) NULL,
       "Print the help."},
      {(char*) NULL, vpParseArgv::ARGV_END, (char*) NULL, (char*) NULL, (char*) NULL}
    };
    // Read the command line options
    if(vpParseArgv::parse(&argc, argv, argTable,
                          vpParseArgv::ARGV_NO_LEFTOVERS |
                          vpParseArgv::ARGV_NO_ABBREV |
                          vpParseArgv::ARGV_NO_DEFAULTS)) {
      return (false);
    }

    if (opt_ipath_)
      opt_ipath = std::string(opt_ipath_);
    if (opt_ppath_)
      opt_ppath = std::string(opt_ppath_);

    return true;
  }
```

The `main()` becomes:

```
    std::string env_ipath;
    std::string opt_ipath;
    std::string ipath;
    std::string opt_ppath;
    std::string ppath;
    std::string filename;
    bool opt_display = true;
    bool opt_click = true;
    float opt_angle = 0;
    float opt_scale_x = 1;
    float opt_scale_y = 1;
    float opt_translation_x = 0;
    float opt_translation_y = 0;

....

    // Read the command line options
    if (getOptions(argc, argv, opt_ipath, opt_ppath, opt_display, opt_click,
                   opt_angle, opt_scale_x, opt_scale_y, opt_translation_x, opt_translation_y) == false) {
      exit (EXIT_FAILURE);
    }

....

    // Make Tansformation Matrix
    T[0][0] =  opt_scale_x * cos(vpMath::rad(opt_angle)); T[0][1] = opt_scale_x * sin(vpMath::rad(opt_angle));  T[0][2] = 0;
    T[1][0] = -opt_scale_y * sin(vpMath::rad(opt_angle)); T[1][1] = opt_scale_y * cos(vpMath::rad(opt_angle));  T[1][2] = 0;
    T[2][0] =  opt_translation_x;                         T[2][1] = opt_translation_y;                          T[2][2] = 1;


```

fspindle replied:
I just found that `ARGV_CONSTANT` can only set a var to `true`. That's why in my previous code, there should be:

```
    bool opt_no_display = false;
    bool opt_no_click = false;
```

and then in the `main()`

```
if(! opt_no_display) // rather then if(opt_display)

```

Same for `opt_no_click`

fspindle commented on modules/core/include/visp3/core/vpImageTools.h:
```diff
@@ -409,6 +420,168 @@ void vpImageTools::crop(const unsigned char *bitmap, unsigned int width, unsigne
+    X[1][j] = B[1];
+    X[2][j] = B[2];
+  }
+
```
Not efficient enough. Instead of doing the SVD for each column, we could compute T inverse just one time and then obtain X = T^-1 * B

The code becomes something like

```
  vpMatrix Tinv = T.inverseByLU();

  // perform T \ X
  for (unsigned int j = 0; j < X.getCols(); j++) {
    B[0] = X[0][j];
    B[1] = X[1][j];
    B[2] = X[2][j];
    vpMatrix X_ = Tinv * B;

    X[0][j] = X_[0][0];
    X[1][j] = X_[1][0];
    X[2][j] = X_[2][0];
  }
```

fspindle commented:
Vikas could you also fill the copyright transfert agreement and send it us back (see [link](https://visp.inria.fr/contributing-code/))

fspindle commented:
1/ I found a strange behavior setting a translation along x + a rotation + scale=2. The left corner of the image is cutted. Should be investigated. Visible running:

`$ ./testImageWarpAffine -angle 45 -trans_x 10 -trans_y 10 -scale_x 2 -scale_y 2` 

2/ There also an other issue that is visible running the test on a 640x480 image. The final size of the image is not the smallest one. Visible running:

`$ ./testImageWarpAffine -angle 45 -trans_x 10 -trans_y 10 -scale_x 2 -scale_y 2`

3/ Check also to be sure that options `-c` and `-d ` are working to disable mouse click and display

vikasTmz commented:
> There also an other issue that is visible running the test on a 640x480 image

I believe this should be fixed with this commit [88af00c](https://github.com/lagadic/visp/pull/152/commits/88af00ca07b7900dd6f328a3438dddb5f5876e02).

About the left corner being clipped, I'm not sure whats causing this, but I'll look into it.

fspindle commented:
There is something going wrong when the transformation is set to Id. In that case input image should be the same as warped image.

Introducing some printings shows that the first pixel is not the same in both images:

````
I[0][0]= (81,81,81,255) I[0][0] warped= (88,88,88,255)
````
If you instrument `vpImageTools::warpAffine()` before calling ` getPixelClamped()` with 
````
  std::cout << "imgx[0][0]: " << imgx[0][0] << std::endl;
  std::cout << "imgy[0][0]: " << imgy[0][0] << std::endl;
````
you will get
````
imgx[0][0]: 1
imgy[0][0]: 1
````
meaning that there is a shift of 1 pixel along the lines and the columns

This is confirmed if you print:

````
    std::cout << "I[0][0]= " << I_image[0][0]
        << " I[1][1]= " << I_image[1][1]
        << " I[0][0] warped= "
        << I_warped_image[0][0] << std::endl;
````
you get:
````
I[0][0]= (81,81,81,255) I[1][1]= (88,88,88,255) I[0][0] warped= (88,88,88,255)
````

This shift could explain the left corner being clipped.

Need to be investigated...

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,250 @@
+  */
+  bool getOptions(int argc, const char **argv, std::string &opt_ipath, std::string &opt_ppath,
+                  bool &opt_no_display, bool &opt_no_click, float &opt_angle,
+                  float &opt_scale_x, float &opt_scale_y,
```
`opt_no_display` and `opt_no_click` should be declared as `int` nor `bool`. I discover that with `bool` it could produce a segfault os OSX.

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,250 @@
+    std::string filename;
+    bool opt_no_display = false;
+    bool opt_no_click = false;
+    // Scale (Sx,Sy), Rotation (angle), Translation (tx,ty)
```
`opt_no_display` and `opt_no_click` should be declared as `int` and initialized to `0`

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,250 @@
+    // Read the command line options
+    if (getOptions(argc, argv, opt_ipath, opt_ppath, opt_no_display, opt_no_click,
+                   opt_angle, opt_scale_x, opt_scale_y, opt_translation_x, opt_translation_y) == false) {
+      exit (EXIT_FAILURE);
```
call `return` instead of `exit`

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,250 @@
+                << "  environment variable to specify the location of the " << std::endl
+                << "  image path where test images are located." << std::endl << std::endl;
+      exit(EXIT_FAILURE);
+    }
```
call `return` instead of `exit`

fspindle commented on modules/core/test/image/testImageWarpAffine.cpp:
```diff
@@ -0,0 +1,250 @@
+
+  std::cout << "\nTest Successful" << std::endl;
+  return EXIT_SUCCESS;
+}
```
I propose the cosmetic changes to ease code understanding:

````
#if defined (VISP_HAVE_X11) || defined (VISP_HAVE_OPENCV) || defined (VISP_HAVE_GTK) || defined (VISP_HAVE_GDI) || defined (VISP_HAVE_D3D9)
    if (!opt_no_display) {
#if defined (VISP_HAVE_X11)
      vpDisplayX *d1 = new vpDisplayX, *d2 = new vpDisplayX;
#elif defined (VISP_HAVE_OPENCV)
      vpDisplayOpenCV *d1 = new vpDisplayOpenCV, *d2 = new vpDisplayOpenCV;
#elif defined (VISP_HAVE_GTK)
      vpDisplayGTK *d1 = new vpDisplayGTK, *d2 = new vpDisplayGTK;
#elif defined (VISP_HAVE_GDI)
      vpDisplayGDI *d1 = new vpDisplayGDI, *d2 = new vpDisplayGDI;
#elif defined (VISP_HAVE_D3D9)
      vpDisplayD3D *d1 = new vpDisplayD3D, *d2 = new vpDisplayD3D;
#else
      std::cerr << "No display available!" << std::endl;
      return EXIT_SUCCESS;
#endif

      d1->init(I_image, 0, 0, "Original image");
      d2->init(I_warped_image, (int) I_image.getWidth()+50, 0, "Warped image");

      vpDisplay::display(I_image);
      vpDisplay::display(I_warped_image);
      vpDisplay::displayText(I_warped_image, 20, 20, "Click to quit.", vpColor::red);
      vpDisplay::flush(I_image);
      vpDisplay::flush(I_warped_image);
      if (!opt_no_click) {
        vpDisplay::getClick(I_warped_image);
      }

      delete d1;
      delete d2;
    }
#endif
````
